### DISCUSSION TITLE
Pass data from _app.getInitialProps to page getInitialProps

### DISCUSSION THREAD

**Discussion by kristojorg** [vercel/next.js#11672] [Help] [answered]
Hi all, I am trying to migrate a large system of websites to Next.js, and I have a quick question. The site is currently using redux to store fetched server data (we will move away from this eventually, but this is how it is for now). I initially build the store in my `_app.getInitialProps` and provide the initial state to the `_app` component, which then passes it down the tree in context. 

However, it is my understanding that tree context is not available in a page's `getInitialProps`. In this function I need to perform some data fetching and update the store with the result. How do you suggest I access the store that I created in `_app`? I have tried passing it directly when calling the page's `getInitialProps` from within `_app`, but despite providing it to the function, it doesn't seem to be passed. See below:
```js
// pages/_app.tsx
function MyApp({
  Component,
  pageProps,
  store
}: AppProps) {
  return (
    <StoreProvider
      store={store}
    >
      <Component {...pageProps} />
    </ContextProvider>
  );
}

MyApp.getInitialProps = async appContext => {
  const store = getOrCreateStore(pathFor);
  /**
   * call the page's getInitialProps, which may further update the redux store.
   */
const pageProps = await App.getInitialProps({
    ...appContext,
  // the store here doesn't actually get passed to getInitialProps
    store,
  });
  /**
   * Pass updated redux state to the app component to be used to rebuild
   * the store on client side with pre-filled data from ssr
   */
  const initialState = store.getState();

  return {
    initialState,
    pageProps
  };
};
```

I have looked at the examples, but they don't seem to initialize the store in the `_app` component, so I don't think they fit my use case.

I also will be moving away from `getInitialProps` to `getServerSideProps` as soon as I can, but I'm not sure how that is possible for my setup currently.

**Comment by eivindml** [comment]
Also interested.

**Comment by DumbledoreD** [comment, marked as answer]
I have a similar setup and my solution is to call `Component.getInitialProps` in `MyApp.getInitialProps`:

```js
// _app.js
MyApp.getInitialProps = async context => {
  const store = initializeStore();
  const componentProps = await context.Component.getInitialProps({ ...context, store });
  return { props: { initialReduxState: store.getState(), ...componentProps }}
};

// _page.js
Page.getInitialProps = async ({ store }) => {
  // ...
}
```

**Reply by kristojorg** [reply]
I was looking for a way to pass the store though. I can pass the state, but I would like to be able to dispatch actions (eg to fetch some data) from within my page's `getInitialProps`. 

Thanks for the input, though!

**Reply by DumbledoreD** [reply]
Sorry, I made a typo in the snippet. Corrected it to `Page.getInitialProps` in the second block. The `store` is passed to the page component's  `getInitialProps`. Make sure your data fetching action returns a promise and resolves after the receive data action is dispatched.

**Reply by kristojorg** [reply]
Ah whoops I was looking at the `initialReduxState` you were passing but that was the props. I am pretty sure I tried this but can't remember why it didn't work for me now. Anyways, thanks for the help!

**Reply by jmndao** [reply]
Hi, thanks for the reply. However, I'm getting an error using this with typescript. Do you have any workaround ?

**Comment by goranefbl** [comment]
In my case I added prop to ctx in _app.js:

appContext.ctx.customProp = 'value';
const appProps = await App.getInitialProps(appContext);

then pull it from ctx in custom page.
